### ISSUE TITLE
a11y: Keyboard method to activate OK button in Change your comment dialog

### ISSUE THREAD

**Issue by jcsteh** [mumble-voip/mumble#2294]
In the "Change your comment" dialog, there's no way to activate the OK buttn using the keyboard, as the text box consumes tab, shift+tab and enter. This is particularly problematic for blind screen reader users who usually rely solely on the keyboard.

One way to solve this is to introduce a key such as control+enter to activate the OK button/accept the change and dismiss the dialog.

**Comment by mkrautz** [comment]
Note to self:

Ctrl+Tab doesn't seem to work here. Ctrl+Tab in the comment dialog's text field switches the tab bar between the "Display" tab and "Source Text" tab.

**Comment by jcsteh** [comment]
> Ctrl+Tab doesn't seem to work here. Ctrl+Tab in the comment dialog's text field switches the tab bar between the "Display" tab and "Source Text" tab.

Which, btw, is the expected behaviour for control+tab; I wouldn't override that.

**Comment by mkrautz** [comment]
OK. Thanks for the clarification!

In the issue for the text message field on the MainWindow, you mention that F6 is often used to switch focus as well.

In this particular case (a dialog box with input), would it make most sense to make it possible to navigate everything with keyboard focus. (I assume yes?).

Or to add Ctrl-Enter to quickly accept?

I assume... having both would be best?

Ctrl-Enter for sending makes a lot of sense in general, for keyboard power users as well.

**Comment by jcsteh** [comment]
F6 isn't standardised as well as, say, the tab key. Where it is 
implemented, f6 swiches between just a few main panes or areas of a 
window; i.e. it shouldn't just be another tab key moving through all 
controls. As an example, try hitting f6 in most web browsers: you'll 
notice it switches between the address bar, the document and maybe 
another important panel. The key point is that there should only be 
maybe two or three things in the "f6 order".

This case is tricky because it's a dialog. I've never seen f6 used in a 
dialog.

I think control+enter for OK is a good solution here and that would be 
enough. However, some might want to move focus to the other controls, so 
bonus points (but far lower priority!) if shift+tab gets you out of the 
input area. :)

**Comment by mkrautz** [comment]
Is it the norm to have to "navigate back" using Shift-Tab to get to controls that are actually *after* a text input field? What would be the solution for such a case?

**Comment by jcsteh** [comment]
It's tricky because I don't think I've ever seen a dialog where an 
editor ate the tab key. I've see this in plenty of full window, 
non-dialog situations (e.g. email compose windows), and in that case, 
having to shift+tab is normal. If it really does need to eat the tab 
key, I think shift+tab to get to the other controls is the best you can 
do. In most cases, people don't need to get to the other controls 
anyway, so they'll normally just hit control+enter (or escape to cancel).

**Comment by mkrautz** [comment]
The "Display" text view doesn't really need to support tab inputs.
The "Source Text" *might*, but still, not necessarily.

I'm curious: is there a standard keyboard shortcut for inserting a tab into a text input field that doesn't allow tab? Just like some input fields allow "Ctrl-Enter" or "Shift-Enter" to force a line break -- even though Enter would normally "Send" the text or accept the dialog or action.

**Comment by jcsteh** [comment]
There's no standard for tab insertion that isn't the tab key, no.

**Comment by Kissaki** [comment]
If TAB is captured, I also expect SHIFT+TAB to be captured.

I would also expect ALT+O for OK, and ALT+C for cancel to work, which it does not.

CTRL+Enter and CTRL+ESC is feasible, and what I would expect.

I was never aware of F6 having the mentioned intention of switching top level focus areas. For me, it was just "focus browser address input".

I’m really wondering now, and trying to remember any application that captures TAB inputs in a dialog…

**Comment by jcsteh** [comment]
> If TAB is captured, I also expect SHIFT+TAB to be captured.

Symmetry is nice, but if it's necessary to capture tab (e.g. for entry of a tab character or auto-completion), that makes things tricky. However, there's no real justification for capturing shift+tab; no one auto-completes or inserts a tab character using shift+tab. Note that this tab/shift+tab asymmetry occurs in the compose area of some email clients such as Mozilla Thunderbird, so this is not "new".

> I would also expect ALT+O for OK, and ALT+C for cancel to work, which it does not.

I've never seen a dialog where these had accelerators, but it's certainly not a bad thing if they do.

> CTRL+Enter and CTRL+ESC is feasible,

Escape already works and is standard for Cancel. Control+escape can't be used because it is often reserved for the OS; e.g. in Windows, it activates the Start Menu.

> I was never aware of F6 having the mentioned intention of switching top level focus areas.

Try it in Windows Explorer, Thunderbird and Microsoft HTML Help, just to name a few. It's possibly only a convention on Windows, though.

> I’m really wondering now, and trying to remember any application that captures TAB inputs in a dialog…

In a dialog, I can't think of any example of this.

**Comment by Kissaki** [comment]
> However, there's no real justification for capturing shift+tab; no one auto-completes or inserts a tab character using shift+tab.

Text and code editors use SHIFT+TAB to remove the tab, to lower indentation.

Thunderbird is a good example. However, SHIFT+Tab is "backwards", which makes sense in Thunderbirds case, because the controls to be focused are above the textarea, and the textarea is the last element - which you can then send with CTRL+Enter.

So there may be use cases for asymmetric TAB/SHIFT+TAB. (TAB as indentation or as textual character) But I am not sure we have a clear candidate here. As we have a text and a source tab, maybe it should be asymmetric in text, and symmetric as indentation in the source tab - which makes it even more confusing.

I would only bind CTRL+Enter for OK.

**Comment by mkrautz** [comment]
Finally found some time to look into this some more...

There is already code in TextMessage.cpp to accept the dialog when Enter is pressed:
https://github.com/mumble-voip/mumble/blob/master/src/mumble/TextMessage.cpp#L64-L76

I checked, and this hasn't worked for a while in wither Qt 4 or Qt 5. Maybe it never worked, or got broken during a refactoring.

Anyway, the problem is that we don't get KeyPress events in the dialog. They're seemingly being swallowed by the text area.

Looking into it.

**Comment by mkrautz** [comment]
Ctrl-Enter behavior is implemented in PR https://github.com/mumble-voip/mumble/issues/2408

I just tested on OS X, and there, Shift-Tab works OK in the dialog to focus the previous item, even when inside the text edit widget. The original report mentions that this *doesn't* work, but it does for me on OS X. Will test on Windows...

**Comment by mkrautz** [comment]
- [ ] Check Shift-Tab behavior in the TextMessage dialog on Windows.

**Comment by Kissaki** [comment]
> Check Shift-Tab behavior in the TextMessage dialog on Windows.

It inserts a tab. So the behaviour is different from OSX, according to your pre-previous comment.

**Comment by Kissaki** [comment]
@mkrautz Did you want to do something in that case?

**Comment by mkrautz** [comment]
OK. We probably want to handle the Backtab event in that case...
